diff --git a/python/sglang/srt/model_loader/loader.py b/python/sglang/srt/model_loader/loader.py
index e2c6a37..4ee6347 100644
--- a/python/sglang/srt/model_loader/loader.py
+++ b/python/sglang/srt/model_loader/loader.py
@@ -653,6 +653,11 @@ class ShardedStateLoader(BaseModelLoader):
                         state_dict.pop(key)
             if state_dict:
                 raise ValueError(f"Missing keys {tuple(state_dict)} in loaded state!")
+
+        if hasattr(model, "post_load_weights"):
+            print("Post loading weights")
+            model.post_load_weights()
+
         return model.eval()
 
     @staticmethod
